File: scripts/descargar-imagenes.mjs
/**
 * Descarga a public/ todas las imagenes /wp-content/uploads/ que referencian
 * los JSON de src/content/pages/, conservando la estructura ano/mes y SIN
 * renombrar nada (estan indexadas en Google Imagenes).
 *
 * Se ejecuta en tu PC, que si tiene salida a prefabricadascasas.es:
 *
 *     node scripts/descargar-imagenes.mjs            (descarga)
 *     node scripts/descargar-imagenes.mjs --ensayo   (solo cuenta)
 *
 * Es idempotente: lo que ya esta descargado no se vuelve a pedir, asi que
 * puedes cortarlo con Ctrl+C y relanzarlo las veces que haga falta.
 */
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const RAIZ = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..');
const PUBLICO = path.join(RAIZ, 'public');
const BASE = 'https://prefabricadascasas.es';

const ensayo = process.argv.includes('--ensayo');
const HILOS = Number(process.argv[process.argv.indexOf('--hilos') + 1]) || 12;

const FIRMAS = [
  [[0xff, 0xd8, 0xff], 'jpg'],
  [[0x89, 0x50, 0x4e, 0x47], 'png'],
  [[0x52, 0x49, 0x46, 0x46], 'webp'],
  [[0x47, 0x49, 0x46, 0x38], 'gif'],
];

function rutasReferenciadas() {
  const dir = path.join(RAIZ, 'src', 'content', 'pages');
  const rutas = new Set();
  const re = /\/wp-content\/uploads\/\d{4}\/\d{2}\/[^\s"'<>)]+?\.(?:jpg|jpeg|png|webp|gif|svg|mp4)/gi;
  for (const f of fs.readdirSync(dir)) {
    if (!f.endsWith('.json')) continue;
    const crudo = fs.readFileSync(path.join(dir, f), 'utf8');
    for (const m of crudo.matchAll(re)) rutas.add(m[0]);
  }
  return [...rutas].sort();
}

function firmaOk(buf, ruta) {
  const ext = ruta.split('.').pop().toLowerCase();
  if (!['jpg', 'jpeg', 'png', 'webp', 'gif'].includes(ext)) return true;
  return FIRMAS.some(([f]) => f.every((b, i) => buf[i] === b));
}

async function descargar(ruta) {
  const destino = path.join(PUBLICO, ruta.replace(/^\//, ''));
  if (fs.existsSync(destino) && fs.statSync(destino).size > 0) return 'ya';

  const url = BASE + ruta.split('/').map(encodeURIComponent).join('/').replace(/%2F/g, '/');
  for (let intento = 0; intento < 3; intento++) {
    try {
      const r = await fetch(url, {
        headers: { 'User-Agent': 'Mozilla/5.0 (migracion prefabricadascasas.es a Astro)' },
        signal: AbortSignal.timeout(30000),
      });
      if (r.status === 404) return '404';
      if (!r.ok) throw new Error('HTTP ' + r.status);
      const buf = Buffer.from(await r.arrayBuffer());
      if (!buf.length) return 'vacio';
      if (!firmaOk(buf, ruta)) return 'firma';
      fs.mkdirSync(path.dirname(destino), { recursive: true });
      fs.writeFileSync(destino, buf);
      return 'ok';
    } catch {
      await new Promise((r) => setTimeout(r, 800 * (intento + 1)));
    }
  }
  return 'error';
}

const rutas = rutasReferenciadas();
const faltan = rutas.filter((r) => !fs.existsSync(path.join(PUBLICO, r.replace(/^\//, ''))));
console.log(`rutas referenciadas: ${rutas.length}`);
console.log(`ya en public/: ${rutas.length - faltan.length} | por descargar: ${faltan.length}`);

if (ensayo) {
  faltan.slice(0, 10).forEach((r) => console.log('   ', r));
  process.exit(0);
}

const cuenta = {};
const fallos = [];
let hecho = 0;

async function trabajador(cola) {
  while (cola.length) {
    const ruta = cola.pop();
    const res = await descargar(ruta);
    cuenta[res] = (cuenta[res] || 0) + 1;
    if (!['ok', 'ya'].includes(res)) fallos.push(`${res}\t${ruta}`);
    if (++hecho % 100 === 0) console.log(`   ${hecho}/${faltan.length}`);
  }
}

const cola = [...faltan];
await Promise.all(Array.from({ length: HILOS }, () => trabajador(cola)));

console.log('\nRESULTADO:', cuenta);
if (fallos.length) {
  fs.writeFileSync(path.join(RAIZ, 'imagenes-fallidas.txt'), fallos.join('\n'));
  console.log(`fallos anotados en imagenes-fallidas.txt (${fallos.length})`);
  fallos.slice(0, 15).forEach((f) => console.log('   ', f));
} else {
  console.log('sin fallos: estan las', rutas.length);
}

File: scripts/descargar-imagenes.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Descarga a public/ todas las imagenes /wp-content/uploads/ referenciadas en
los JSON de src/content/pages/, conservando la estructura ano/mes y SIN
renombrar nada (estan indexadas en Google Imagenes).

Uso: python3 scripts/descargar-imagenes.py [--ensayo] [--hilos N]
"""
import os, re, sys, json, glob, time
from concurrent.futures import ThreadPoolExecutor
import urllib.request, urllib.error

BASE = "https://prefabricadascasas.es"
RAIZ = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
PUBLICO = os.path.join(RAIZ, "public")

FIRMAS = {
    b"\xff\xd8\xff": "jpg",
    b"\x89PNG": "png",
    b"RIFF": "webp",
    b"GIF8": "gif",
}

def rutas_referenciadas():
    rutas = set()
    for f in glob.glob(os.path.join(RAIZ, "src", "content", "pages", "*.json")):
        d = json.load(open(f, encoding="utf-8"))
        texto = d.get("cuerpo", "") + json.dumps(d.get("hero", {}), ensure_ascii=False)
        rutas |= set(re.findall(
            r"/wp-content/uploads/\d{4}/\d{2}/[^\s\"'<>)]+?\.(?:jpg|jpeg|png|webp|gif|svg|mp4)",
            texto, re.I))
    return sorted(rutas)

def descargar(ruta):
    destino = os.path.join(PUBLICO, ruta.lstrip("/"))
    if os.path.exists(destino) and os.path.getsize(destino) > 0:
        return ("ya", ruta, os.path.getsize(destino))
    os.makedirs(os.path.dirname(destino), exist_ok=True)
    url = BASE + urllib.parse.quote(ruta, safe="/-_.~")
    for intento in range(3):
        try:
            pet = urllib.request.Request(url, headers={
                "User-Agent": "Mozilla/5.0 (migracion prefabricadascasas.es a Astro)"})
            with urllib.request.urlopen(pet, timeout=30) as r:
                datos = r.read()
            if not datos:
                return ("vacio", ruta, 0)
            # verificacion de firma (svg y mp4 no la tienen)
            ext = ruta.rsplit(".", 1)[-1].lower()
            if ext in ("jpg", "jpeg", "png", "webp", "gif"):
                if not any(datos.startswith(f) for f in FIRMAS):
                    return ("firma", ruta, len(datos))
            with open(destino, "wb") as f:
                f.write(datos)
            return ("ok", ruta, len(datos))
        except urllib.error.HTTPError as e:
            if e.code == 404:
                return ("404", ruta, 0)
            time.sleep(1 + intento)
        except Exception:
            time.sleep(1 + intento)
    return ("error", ruta, 0)

def main():
    import urllib.parse  # noqa
    ensayo = "--ensayo" in sys.argv
    hilos = 12
    if "--hilos" in sys.argv:
        hilos = int(sys.argv[sys.argv.index("--hilos") + 1])

    rutas = rutas_referenciadas()
    print("rutas referenciadas: %d" % len(rutas))
    faltan = [r for r in rutas if not os.path.exists(os.path.join(PUBLICO, r.lstrip("/")))]
    print("ya en public/: %d | por descargar: %d" % (len(rutas) - len(faltan), len(faltan)))
    if ensayo:
        for r in faltan[:10]:
            print("   ", r)
        return

    resultados = []
    hecho = 0
    with ThreadPoolExecutor(max_workers=hilos) as ex:
        for res in ex.map(descargar, faltan):
            resultados.append(res)
            hecho += 1
            if hecho % 200 == 0:
                print("   %d/%d" % (hecho, len(faltan)), flush=True)

    from collections import Counter
    cuenta = Counter(r[0] for r in resultados)
    bytes_ok = sum(r[2] for r in resultados if r[0] == "ok")
    print("\nRESULTADO:", dict(cuenta))
    print("descargado: %.1f MB" % (bytes_ok / 1e6))
    fallos = [r for r in resultados if r[0] not in ("ok", "ya")]
    if fallos:
        with open(os.path.join(RAIZ, "imagenes-fallidas.txt"), "w") as f:
            f.write("\n".join("%s\t%s" % (a, b) for a, b, _ in fallos))
        print("fallos anotados en imagenes-fallidas.txt (%d)" % len(fallos))
        for a, b, _ in fallos[:15]:
            print("   %-6s %s" % (a, b))

if __name__ == "__main__":
    import urllib.parse
    main()
